### PR TITLE
fix: improve generic quote evasion detection

### LINKED ISSUE #3397: Natural language FPs with rule 932240 PL2
_This issue is part of the ongoing_ quantitative testing _natural language FPs improvement task. See the main issue for details: #3392_

### Description

Rule 932240 struggles with false positives on natural language input. This is because it matches on apostrophes and quote marks in the middle of sentences (either single `'` or double `"`).

This rule, 932240, is one of the new Unix RCE rules:
`Remote Command Execution: Unix Command Injection evasion attempt detected`
`(generic detection, PL2, targets generic evasion attempts)`

I believe this is the rule that tries to replicate (and improve upon) ModSecurity's `cmdline` transformation function. This rule is designed to detect evasion-like patterns:

```
- non-existing vars: cu$@rl, una$$me -a, uname$$u -a
- vars + spacing: v='u';cu$v\r\l google.com
- globbing pattern expansion: {n$u\c$u,-nlvp,777}
- globbing: garb=cur[l];$garb+google.com
```

### Logs

```
Number	Payload
300	All told, people were taken with the concept of Stadia, but the execution just didn't work.
Rules triggered
  [Thu Nov 30 00:08:18.160748 2023]
      ModSecurity: Warning. Match of "rx [0-9]\\\\s*\\\\'\\\\s*[0-9]" against "MATCHED_VAR" required.
      [file "/etc/modsecurity.d/owasp-crs/rules/REQUEST-932-APPLICATION-ATTACK-RCE.conf"]
      [id "932240"]
      [msg "Remote Command Execution: Unix Command Injection evasion attempt detected"]
      [data "Matched Data: didn't found within MATCHED_VAR: All told, people were taken with the concept of Stadia, but the execution just didn't work."]
      [severity "CRITICAL"]
      [ver "OWASP_CRS/4.0.0-rc2"]
      ["paranoia-level/2"]
      [unique_id "ZWfSchXIgioxuQ8PfmG0PwAAAAA"]
```

```
Number	Payload
1430	But three years before that movie came out, he appeared in the episode "Bad Blood" of as Sheriff Hartwell.
Rules triggered
  [Thu Nov 30 00:29:31.637490 2023]
      ModSecurity: Warning. Match of "rx [0-9]\\\\s*\\\\'\\\\s*[0-9]" against "MATCHED_VAR" required.
      [file "/etc/modsecurity.d/owasp-crs/rules/REQUEST-932-APPLICATION-ATTACK-RCE.conf"]
      [id "932240"]
      [msg "Remote Command Execution: Unix Command Injection evasion attempt detected"]
      [data "Matched Data: Blood\\x22 of found within MATCHED_VAR: But three years before that movie came out, he appeared in the episode \\x22Bad Blood\\x22 of as Sheriff Hartwell."]
      [severity "CRITICAL"]
      [ver "OWASP_CRS/4.0.0-rc2"]
      ["paranoia-level/2"]
      [unique_id "ZWfXa2kRdL2E9hp7KX-dcQAAAIU"]
```

```
Number	Payload
5270	Now it's the older and more seasoned sprinters' turn.
Rules triggered
  [Thu Nov 30 00:29:39.646090 2023]
      ModSecurity: Warning. Match of "rx [0-9]\\\\s*\\\\'\\\\s*[0-9]" against "MATCHED_VAR" required.
      [file "/etc/modsecurity.d/owasp-crs/rules/REQUEST-932-APPLICATION-ATTACK-RCE.conf"]
      [id "932240"]
      [msg "Remote Command Execution: Unix Command Injection evasion attempt detected"]
      [data "Matched Data: it's found within MATCHED_VAR: Now it's the older and more seasoned sprinters' turn."]
      [severity "CRITICAL"]
      [ver "OWASP_CRS/4.0.0-rc2"]
      ["paranoia-level/2"]
      [unique_id "ZWfXcxXIgioxuQ8PfmG1GwAAAAQ"]
```

### How to reproduce the misbehavior (-> curl call)

```
curl https://sandbox.coreruleset.org -H 'x-format-output: txt-matched-rules' -H 'x-crs-paranoia-level: 2' --data "payload=All told, people were taken with the concept of Stadia, but the execution just didn't work."
```

### Your Environment

* CRS version (e.g., v3.3.4): v4.0.0-rc2
* Paranoia level setting (e.g. PL1) : PL 2
* Operating System and version: CRS modsec2-apache container

### Confirmation

[x] I have removed any personal data (email addresses, IP addresses,
    passwords, domain names) from any logs posted.

### PR DESCRIPTION
**Pull request by theseion**
- do not detect single quotes as evasion
- detect multiple single or double quotes as evasion attempt
- detect quoted content like `kill '-'9`

Fixes #3397

### CODE REVIEW

**Comment by RedXanadu** [comment]
**Before:**

```
Freq.   ID #   Paranoia Level
    822 932240 PL2
```

**After (this PR):**

```
Freq.   ID #   Paranoia Level
    229 932240 PL2
```

**Comment by theseion** [comment]
Thanks. Still not perfect but much better at least.

**Comment by dune73** [comment]
Thanks. I get the feeling this is acceptable for a release, but let's ask the team.

**Comment by franbuehler** [comment]
I think this is a good solution to test for `'` _inside_ a word to detect evasions. I tested the new regex and it's better that the old one.

- `hello'world'xx` -> match
- `hello'world' xx` -> no match
- `pyth''on` -> match
- `hello 'world'x` -> match
- `hello 'world'` -> no match

**Review by franbuehler — APPROVED** [review]

**Comment by theseion** [comment]
I've improved on the detection of things like `kill '-'9` to fix the FPs reported by @xanadu.

**Comment by RedXanadu** [comment]
@theseion Awesome! Can we chalk this one up as a victory? :man_dancing:

| Freq.  | ID #   | Paranoia Level |
| ------ | ------ | -------------- |
|     11 | 932240 | PL2            |

**Comment by theseion** [comment]
I sure hope so :) I just need an approving review from somebody.

**Comment by dune73** [comment]
If that's not a win, then I do not know!

It looked horrible in the beginning and now it's a homerun.

File: tests/regression/tests/REQUEST-932-APPLICATION-ATTACK-RCE/932240.yaml [hunk 1]
```diff
@@ -264,3 +264,75 @@ tests:
...
+              Host: "localhost"
+              Accept: text/xml,application/xml,application/xhtml+xml,text/html;q=0.9,text/plain;q=0.8,image/png,*/*;q=0.5
+            uri: "/get"
+            data: |
+              payload=kill '-'9
+            protocol: "http"
+          output:
+            log_contains: id "932240"
+  - test_title: 932240-18
+    desc: "False positive against quote evatsio with two quotes"
```
Comment (azurit) [inline tests/regression/tests/REQUEST-932-APPLICATION-ATTACK-RCE/932240.yaml:304]: Typo in `evatsio`.

**Review by azurit — APPROVED** [review]

**Comment by azurit** [comment]
Thank you, good work!

**Pull request merged**